### PR TITLE
Filter in DataGrid column headers jmix-framework/jmix#1752

### PR DESCRIPTION
**Pull request by KremnevDmitry**
Related to: jmix-framework/jmix#1752

### CODE REVIEW

File: jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/element/StudioElements.java [hunk 1]
```diff
@@ -91,6 +91,8 @@ public interface StudioElements {
             properties = {
                     @StudioProperty(xmlAttribute = "autoWidth", type = StudioPropertyType.BOOLEAN,
                             defaultValue = "false"),
+                    @StudioProperty(xmlAttribute = "filterable", type = StudioPropertyType.BOOLEAN,
+                            defaultValueRef = "parent:filterable"),
```
Comment (KremnevDmitry) [inline jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/element/StudioElements.java:95]: Interesting moment
See: https://github.com/jmix-framework/jmix/issues/2128

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java [hunk 2]
```diff
@@ -362,6 +383,157 @@ protected Optional<Renderer> loadRenderer(Element columnElement, @Nullable MetaP
...
+            }
+        });
+
+        return filterButton;
+    }
+
+    @SuppressWarnings({"rawtypes"})
+    protected Dialog createOverlay(PropertyFilter propertyFilter, JmixButton filterButton) {
+        Dialog dialog = new Dialog(propertyFilter);
+        dialog.addClassName(COLUMN_FILTER_DIALOG_CLASSNAME);
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:451]: what if we set `themeName`?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java [hunk 3]
```diff
@@ -362,6 +383,157 @@ protected Optional<Renderer> loadRenderer(Element columnElement, @Nullable MetaP
...
+        }
+
+        AtomicReference appliedValue = new AtomicReference<>();
+
+        dialog.getFooter().add(
+                createApplyButton(propertyFilter, dialog, appliedValue),
+                createCancelButton(propertyFilter, dialog, appliedValue)
+        );
+
+        dialog.addOpenedChangeListener(event -> {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:466]: extract lambda to method?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java [hunk 4]
```diff
@@ -362,6 +383,157 @@ protected Optional<Renderer> loadRenderer(Element columnElement, @Nullable MetaP
...
+
+        dialog.close();
+    }
+
+    protected Component createHeaderComponent(String headerText, Component filterButton) {
+        HorizontalLayout layout = getUiComponents().create(HorizontalLayout.class);
+        layout.setPadding(false);
+        layout.setSpacing(false);
+        layout.setClassName(LumoUtility.Gap.XSMALL);
+        layout.getStyle().set("padding", "2px");
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:531]: why?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java [hunk 5]
```diff
@@ -137,32 +154,31 @@ protected void loadMultiSort() {
...
-            case COLUMN_ELEMENT_NAME:
-                loadColumn(resultComponent, columnElement, metaClass);
-                break;
-            case EDITOR_ACTIONS_COLUMN_ELEMENT_NAME:
-                loadEditorActionsColumn(resultComponent, columnElement);
-                break;
-            default:
-                throw new GuiDevelopmentException("Unknown columns' child element: " + columnElement.getName(),
-                        context, "Component ID", resultComponent.getId());
+            case COLUMN_ELEMENT_NAME -> loadColumn(resultComponent, columnElement, holder.getDataLoader(),
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:177]: We don't refactor code just to refactor it.

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java (hunk 2 above)
Comment (KremnevDmitry) [reply jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:451]: No)

Comment (glebfox) [reply jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/AbstractGridLoader.java:456]: image

**Review by glebfox — APPROVED** [review]

**Pull request merged**